### PR TITLE
Hyper upgrade

### PR DESCRIPTION
**Pull request by audunhalland**
My attempt at upgrading hyper to a `futures`-aware version.

Pact testing utils for rust also now futures-aware (see tests in `pact_broker.rs`).

I have not upgraded hyper in the `pact_mock_server_cli` crate. I'm proposing intruction of a more commonly used web framework there, I suggest https://github.com/seanmonstar/warp, but I feel it is outside the scope of this PR, as it's getting big enough.

### CODE REVIEW

File: rust/pact_consumer/src/mock_server.rs [hunk 1]
```diff
@@ -32,26 +51,53 @@ impl StartMockServer for Pact {
...
     // The URL of our mock server.
     url: Url,
+    // The mock server instance
+    mock_server: mock_server::MockServer,
+    // The running mode of our mock server.
+    _mode: Mode
 }
 
 impl ValidatingMockServer {
+    /// Create a new mock server
```
Comment (audunhalland) [inline rust/pact_consumer/src/mock_server.rs]: I think there should be tests here for both of the running modes of the validating mock server, and perhaps better documentation overall.

The reason there is two running modes is as you may guess the ability to test both synchronous and asynchronous http clients. When http clients under test are async, tests are more performant when running the mock server without allocating any extra threads.

**Comment by uglyog** [comment]
Wow, awesome work. I know what a big effort this is! I'll go through it over the next few days.

We definitely can look to move to a better web framework now that there is more crates available. But I agree, let us get this one over the line first.

File: rust/pact_mock_server/src/hyper_server.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+    body.to_string()
+}
+
+fn match_result_to_hyper_response(request: &Request, match_result: MatchResult) -> Result<Response<Body>, InteractionError> {
+    match match_result {
+        MatchResult::RequestMatch(ref interaction) => {
+            let response = pact_matching::generate_response(&interaction.response);
+            info!("Request matched, sending response {:?}", response);
+            info!("     body: '{}'\n\n", interaction.response.body.str_value());
+            info!("     body: '{}'\n\n", interaction.response.body.str_value());
```
Comment (uglyog) [inline rust/pact_mock_server/src/hyper_server.rs]: Duplicated line

File: rust/pact_mock_server/src/hyper_server.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+        Ok(response) => Ok(response),
+        Err(error) => {
+            let response = match error {
+                InteractionError::RequestHeaderEncodingError => Response::builder()
+                    .status(400)
+                    .body(Body::from("Found an invalid header encoding")),
+                InteractionError::RequestBodyError => Response::builder()
+                    .status(500)
+                    .body(Body::from("Could not process request body")),
+                InteractionError::ResponseBodyError => Response::builder()
```
Comment (uglyog) [inline rust/pact_mock_server/src/hyper_server.rs]: Duplicated case

File: rust/pact_mock_server_cli/src/server.rs [hunk 4]
```diff
@@ -179,20 +183,27 @@ fn shutdown_resource(server_key: Arc<String>) -> WebmachineResource {
   }
 }
 
-fn mock_server_resource(output_path: Arc<Option<String>>) -> WebmachineResource {
+fn mock_server_resource(
+    output_path: Arc<Option<String>>,
+    server_manager: Arc<Mutex<ServerManager>>
+) -> WebmachineResource {
+    let server_manager1 = server_manager.clone();
```
Comment (uglyog) [inline rust/pact_mock_server_cli/src/server.rs]: This is one of the things I hate about Rust! :-D

File: rust/pact_mock_server/src/hyper_server.rs (hunk 3 above)
Comment (uglyog) [reply rust/pact_mock_server/src/hyper_server.rs]: Oh, wait, request and response errors. No issue here.

File: rust/pact_mock_server_cli/src/server.rs (hunk 4 above)
Comment (audunhalland) [reply rust/pact_mock_server_cli/src/server.rs]: Yes, it's not nice. The pattern does tend show up in closure-heavy code, which often makes me covert closures to functions, but I did not want to do a big rewrite here.

**Comment by audunhalland** [comment]
With `ServerManager` in place, only the FFI-type library functions mutate any global state. This is an indication to me that this part of the library could be factored out into a separate, thin "ffi" crate, but it's not too big of a deal.

**Comment by uglyog** [comment]
A separate thin "ffi" crate is a good idea.

**Pull request merged**